### PR TITLE
Folia Support

### PR DESCRIPTION
**Pull request by YufiriaMazenta**
Added Folia compatibility for plugin
https://github.com/kernitus/BukkitOldCombatMechanics/issues/708

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -72,7 +76,7 @@ description = "OldCombatMechanics"
 java {
     toolchain {
         // At least 17 required for MC 1.19 for ingametesting
-        languageVersion.set(JavaLanguageVersion.of(8))
+        languageVersion.set(JavaLanguageVersion.of(17))
```
Comment (kernitus) [inline build.gradle.kts:79]: We can't bump the minimum java version without breaking compatibility with people using older versions e.g. 1.12

File: src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/BukkitScheduler.java
Comment (kernitus) [inline src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/BukkitScheduler.java:1]: Why is a whole separate class needed that just wraps other methods? Does folia not provide a better way of doing this?

File: src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/SchedulerManager.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package kernitus.plugin.OldCombatMechanics.scheduler;
+
+public enum SchedulerManager {
+
+    INSTANCE;
+    private IScheduler scheduler;
+
+    public void loadPlatform() {
+        try {
+            Class<?> pluginMetaClass = Class.forName("io.papermc.paper.plugin.configuration.PluginMeta");
```
Comment (kernitus) [inline src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/SchedulerManager.java:10]: Is there no public API to figure this out?

File: src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/IScheduler.java
Comment (kernitus) [inline src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/IScheduler.java:1]: Could you please translate the chinese comments into english?

**Review by kernitus — CHANGES_REQUESTED** [review]
Is there an actual guide with examples on steps needed to support Folia? I couldn't find anything in their documentation

File: src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/BukkitScheduler.java
Comment (YufiriaMazenta) [reply src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/BukkitScheduler.java:1]: Folia does not provide a unified API. The method it provides to determine whether it is the folia only exists on the paper above 1.19.4, so there is currently no better way. The purpose of my separate encapsulation is for versatility.

File: src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/SchedulerManager.java (hunk 2 above)
Comment (YufiriaMazenta) [reply src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/SchedulerManager.java:10]: Currently folia related api are judged through reflection, such as UniversalScheduler

**Comment by YufiriaMazenta** [comment]
> 是否有实际指南，其中包含支持 Folia 所需步骤的示例？我在他们的文档中找不到任何内容

The adaptation on the folia side mainly lies in thread-related calls, and other APIs are basically implemented in the same way I do.

**Comment by YufiriaMazenta** [comment]
> Is there an actual guide with examples on steps needed to support Folia? I couldn't find anything in their documentation

The adaptation on the folia side mainly lies in thread-related calls, and other APIs are basically implemented in the same way I do.

File: build.gradle.kts (hunk 1 above)
Comment (YufiriaMazenta) [reply build.gradle.kts:79]: It seems that using folia-api must use java17 to compile. I don’t know how to solve it for the time being.

Comment (kernitus) [reply build.gradle.kts:78]: Spigot up to 1.20.3-R0.1 uses Java 8. Are there older versions of the Folia API?

Comment (YufiriaMazenta) [reply build.gradle.kts:78]: The minimum folia-api version is 1.19.4

Comment (YufiriaMazenta) [reply build.gradle.kts]: I have modified it to build with java8

File: build.gradle.kts [hunk 3]
```diff
@@ -52,11 +54,21 @@ dependencies {
...
-
-     //For ingametesting
+    // Guava
+    compileOnly("com.google.guava:guava:33.0.0-jre")
+    // Gson
+    compileOnly("com.google.code.gson:gson:2.10.1")
+    // Adventure
+    compileOnly("net.kyori:adventure-api:4.16.0")
+    // BungeeCord Chat
+    compileOnly("net.md-5:bungeecord-chat:1.16-R0.4")
```
Comment (kernitus) [inline build.gradle.kts:69]: What is bungeecord chat needed for?

File: build.gradle.kts [hunk 4]
```diff
@@ -52,11 +54,21 @@ dependencies {
     // For BSON file serialisation
     implementation("org.mongodb:bson:4.11.1")
     // Spigot
-    compileOnly("org.spigotmc:spigot-api:1.20.3-R0.1-SNAPSHOT")
+//    compileOnly("org.spigotmc:spigot-api:1.20.3-R0.1-SNAPSHOT")
+    // Folia
+    compileOnly("dev.folia:folia-api:1.19.4-R0.1-20230527.192927-40@jar")
```
Comment (kernitus) [inline build.gradle.kts:59]: Did you actually manage to get this to build with Java 1.8? The github actions pipeline seems to fails to build this

File: src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/BukkitScheduler.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import kernitus.plugin.OldCombatMechanics.scheduler.task.BukkitTaskWrapper;
+import kernitus.plugin.OldCombatMechanics.scheduler.task.ITaskWrapper;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.entity.Entity;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Bukkit平台的调度器
```
Comment (kernitus) [inline src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/BukkitScheduler.java:12]: Could you translate this to english please

File: src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/FoliaScheduler.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+import org.bukkit.Location;
+import org.bukkit.entity.Entity;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+/**
+ * Folia平台的调度器
```
Comment (kernitus) [inline src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/FoliaScheduler.java:16]: Could you translate this to english please

File: src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/task/ITaskWrapper.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package kernitus.plugin.OldCombatMechanics.scheduler.task;
+
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.NotNull;
+
+public interface ITaskWrapper {
+
+    /**
+     * 取消此任务
```
Comment (kernitus) [inline src/main/java/kernitus/plugin/OldCombatMechanics/scheduler/task/ITaskWrapper.java:9]: Could you translate this to english please

**Review by kernitus — CHANGES_REQUESTED** [review]
Have you build this and tried running it with folia yet?

**Comment by kernitus** [comment]
@YufiriaMazenta is there a reason you closed the PR?

**Comment by ugmapekibert** [comment]
when will it be developed for folia? I ran this plugin on folia and it worked, but after every second run in plugin.yml, the line "folia-supported: true" disappeared, don't know what's the matter?

**Comment by kernitus** [comment]
> when will it be developed for folia? I ran this plugin on folia and it worked, but after every second run in plugin.yml, the line "folia-supported: true" disappeared, don't know what's the matter?

Folia is not currently supported but support could be added after the current rewrite - once we're sure all the main core mechanics of the plugin still work as intended.

**Comment by ugmapekibert** [comment]
> > when will it be developed for folia? I ran this plugin on folia and it worked, but after every second run in plugin.yml, the line "folia-supported: true" disappeared, don't know what's the matter?
> 
> Folia is not currently supported but support could be added after the current rewrite - once we're sure all the main core mechanics of the plugin still work as intended.

thanks

**Pull request closed without merging**